### PR TITLE
feat(telemetry): hook fire-rate ROI ledger

### PR DESCRIPTION
**Pull request by devseunggwan**
## 무엇을 / 왜

ponytail 비교(deep-dive)에서 도출된 P1. praxis는 ~60개 훅을 직관으로 누적했지만 그 훅들이 실제로 마찰을 줄이는지 측정한 적이 없다. 이 PR은 훅 fire-rate를 기록하는 observe-only telemetry 기반을 추가해, praxis가 모든 곳에 요구하는 evidence-based 잣대를 자기 훅에도 처음으로 적용한다. #713(증거 기반 hook prune 감사)의 선행 데이터 기반이다.

## 커버리지 (2-tier)

- **RICH** — 중앙 PreToolUse(Bash) dispatcher(`_dispatch.py`)가 이미 캡처하는 각 멤버의 `(rc, stdout, stderr)`를 `block/ask/advise/pass`로 분류해 full-granularity 기록.
- **COARSE** — 그 외 모든 `@fail_open` standalone 훅(Stop / UserPromptSubmit / PostToolUse / 비-Bash PreToolUse). 유일한 공통 chokepoint인 `@fail_open` 데코레이터를 **비침습적**으로 계측 — `main()` 반환 후 return code만 보고 기록(stdin/stdout/stderr 미접촉, 훅 동작 불변).
  - **block 한계(정직성)**: coarse "block"은 **exit-code 2만** 집계. Stop/UserPromptSubmit 훅은 stdout JSON `decision` 필드로 차단하며 exit 0이라, 그 차단은 coarse에 보이지 않고 "pass"로 기록된다(PreToolUse 차단은 exit 2라 잡힘). docstring·리포트에 명시.
- **이중기록 방지**: dispatcher 프로세스는 `mark_dispatcher_process()`로 자기 표시 → Bash 그룹 멤버는 rich로만 기록, coarse 경로는 skip.
- **shell(.sh) 훅**: `@fail_open`/dispatch chokepoint가 없어 계측 불가 → never-fired roster에서 제외하고 별도 "Uninstrumented" 섹션에 분리(codex #714 P2).

## 변경 (5파일)

- `hooks/_lib/_fire_ledger.py` (신규): observe-only writer. `record_group_fires`(rich, batched) + `record_standalone_fire`(coarse) + `_atomic_append`(per-line `O_APPEND` 원자 추가 + 정규파일 가드/`O_NOFOLLOW`로 FIFO/symlink hang·오도 차단) + granularity 필드. bypass-telemetry 계약(opt-out `PRAXIS_FIRE_TELEMETRY_DISABLE`, override `PRAXIS_FIRE_TELEMETRY_FILE`, fail-open) 미러.
- `hooks/_lib/_hook_runtime.py`: `fail_open`이 `main()` 반환 후 coarse fire 기록(`_maybe_record_fire`). 반환값/예외/I/O 스트림 불변 — fail-open 계약 유지.
- `hooks/_lib/_dispatch.py`: `run_group` 진입 시 dispatcher 프로세스 표시. 결정 로직 불변.
- `skills/bypass-review/bypass-review`: `fire-rate` 모드. per-hook 카운트 + rich/coarse(G) 컬럼 + instrumentable roster 대비 never-fired + uninstrumented(shell) 별도 섹션 + host-blind/rare-but-critical caveat. 기존 호출 back-compat.
- `tests/test_fire_ledger.py`: decision 4종 precedence + standalone coarse 기록 + dispatcher-skip(이중기록 방지) + roster split + fail_open 통합 + FIFO 가드 + writer↔reader roundtrip(SUT-mirror 회피) + back-compat.

## 검증

- 단위 28(파일 단독) / 373(전체 dir) 통과, 전체 스위트(`scripts/run-tests.sh`) **ALL TESTS PASSED**, ruff clean.
- **실제 아티팩트 기능 검증**: 실제 standalone 훅(Stop/PostToolUse) → coarse 기록, 실제 `_dispatch.py` + 32개 Bash-group 훅 → rich 기록, **coarse∩rich 겹침 0**(이중기록 없음). telemetry 파일을 FIFO로 둔 실제 Stop 훅 → **멈춤 없음**(FIFO 가드 동작 확인).
- 3종 독립 리뷰: omc code-reviewer(HIGH 1 = Stop/UPS block-as-pass 정직성 반영, MEDIUM 2 = docstring) + security-reviewer(LOW = FIFO/symlink 가드 반영) + codex(P2 = shell 훅 roster 분리 반영).

## 검증 안 한 것

- 실제 훅의 `ask`/`advise` decision 및 Stop의 stdout-JSON block 경로(coarse는 설계상 미구분 — return code만 관측).
- 병렬 dispatcher 동시 append 부하(원자성은 per-line `O_APPEND`로 설계상 보장, 부하 테스트는 미실시).

## 영향 범위

`@fail_open`은 모든 standalone 훅 + dispatcher의 hot path. 변경은 observe-only·이중 fail-open이며 I/O 스트림·반환값·결정 동작에 영향 없음. telemetry 비활성화 시 무비용 단락.

## Open items (후속)

- 텔레메트리 retention 정책(일별 파일 누적 — reader는 `--days` 윈도).
- coarse의 Stop/UPS block 가시성: stdout 캡처는 훅 간섭 위험으로 의도적 미채택(현재는 정직성 caveat로 처리).

Refs #710 (fire-ledger + 전 훅 fire-rate 리포트 구현; outcome proxy 및 A/B sub-phase는 후속으로 연기 — 이슈 유지)

## Summary by CodeRabbit

* **New Features**
  * Added per-hook “fire” telemetry capture and logging (rich for dispatched groups, coarse for standalone hooks).
  * Introduced a new `fire-rate` report mode with decision breakdowns and a “Never Fired” section when manifest data is available.

* **Bug Fixes**
  * Telemetry recording is now fail-open and won’t block dispatch or reporting, including during malformed/edge cases.
  * Coarse standalone recording is suppressed during dispatcher processing to avoid duplicate telemetry.

* **Tests**
  * Expanded coverage for decision classification, telemetry writing/opt-out behavior, report aggregation, and suppression/fail-open semantics.

### CODE REVIEW

File: hooks/_lib/_fire_ledger.py [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+                "hook": name,
+                "role": role,
+                "decision": classify_decision(rc, stdout, stderr),
+            }, ensure_ascii=False))
+        if not lines:
+            return
+        path = resolve_path()
+        path.parent.mkdir(parents=True, exist_ok=True)
+        with path.open("a", encoding="utf-8") as fh:
+            fh.write("\n".join(lines) + "\n")
```
[the remark this answers is not in the document]
Comment (devseunggwan) [reply hooks/_lib/_fire_ledger.py]: Fixed in bc6f46a — switched to per-line os.write(O_APPEND). Each fire record (~200B) is well under PIPE_BUF, so appends are atomic without a cross-process lock; whole-line interleaving across processes is harmless for line-oriented JSONL. Avoids adding lock latency to the dispatcher hot path.

File: skills/bypass-review/bypass-review [hunk 2]
```diff
@@ -648,6 +681,172 @@ def _render_error_events(lines: list[str], error_events: list[dict]) -> None:
...
+    for hook in manifest.get("hooks", []):
+        name = hook.get("name")
+        if not name:
+            continue
+        entries = hook.get("entries") or [
+            {"event": hook.get("event"), "matcher": hook.get("matcher")}
+        ]
+        for entry in entries:
+            if entry.get("event") == "PreToolUse" and entry.get("matcher") == "Bash":
+                roster.add(name)
```
[the remark this answers is not in the document]
Comment (devseunggwan) [reply skills/bypass-review/bypass-review]: Fixed in d4b09b6 — added isinstance guards on hook/entries plus a full try/except returning None. A valid-JSON-but-misstructured manifest now skips the bad item instead of raising. Regression test added (test_bash_group_roster_malformed_manifest_does_not_crash).

File: skills/bypass-review/bypass-review [hunk 3]
```diff
@@ -648,6 +681,172 @@ def _render_error_events(lines: list[str], error_events: list[dict]) -> None:
...
+
+    lines.append(_render_header("bypass-review: Hook Fire-Rate Report (issue #710)"))
+    lines.append(f"  Period : {since} → {today} (last {days} day{'s' if days != 1 else ''})")
+    lines.append(f"  Source : {telemetry_dir}")
+    lines.append(f"  Hooks fired : {len(by_hook)}")
+    lines.append("  Scope  : PreToolUse(Bash) dispatch group only (hot-path MVP ceiling)")
+
+    if not by_hook:
+        lines.append("\n  No fire events found in the selected period.")
+        return "\n".join(lines)
```
[the remark this answers is not in the document]
Comment (devseunggwan) [reply skills/bypass-review/bypass-review:840]: By design, not changing. When zero events exist, listing the full 32-hook Bash roster as 'never fired' would mislabel 'no telemetry collected yet' as 'dead hooks'. The early return keeps the explicit 'No fire events found' message. The never-fired set is meaningful only once some hooks have fired in the window.

File: hooks/_lib/_fire_ledger.py [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+    """
+    if not lines:
+        return
+    path.parent.mkdir(parents=True, exist_ok=True)
+    try:
+        if not stat.S_ISREG(os.lstat(path).st_mode):
+            return  # FIFO / device / socket / symlink — refuse to write
+    except FileNotFoundError:
+        pass  # absent — created as a regular file below
+    fd = os.open(path, os.O_WRONLY | os.O_APPEND | os.O_CREAT | getattr(os, "O_NOFOLLOW", 0), 0o644)
```
[the remark this answers is not in the document]
Comment (devseunggwan) [reply hooks/_lib/_fire_ledger.py]: Fixed in e803c84 — added O_NONBLOCK (a FIFO swapped in after the lstat now fails open() with ENXIO instead of blocking) and an os.fstat() check on the opened fd before writing, closing the lstat→open swap window. Kept inside the caller's fail-open try/except.

File: hooks/_lib/_hook_runtime.py [hunk 5]
```diff
@@ -110,13 +137,23 @@ def _record_swallowed_exception(fn: Callable[[], int]) -> None:
...
-            return fn()
+            rc = fn()
         except Exception:
             _record_swallowed_exception(fn)
-            return 0
+            rc = 0
+        # Coarse fire telemetry (issue #710 coverage). After fn() resolves, never
+        # alters rc or raises. Skipped inside the dispatcher process.
+        _maybe_record_fire(fn, rc)
+        return rc
```
[the remark this answers is not in the document]
Comment (devseunggwan) [reply hooks/_lib/_hook_runtime.py:163]: Fixed in e803c84 — added an `except SystemExit` clause before the generic Exception handler that records the fire from the exit code and re-raises to preserve exit semantics. Note: an AST scan found 0 hooks that call sys.exit() inside their @fail_open main() today (the convention is `sys.exit(main())`, outside the wrapper), so this is defensive against future drift rather than a current miss.

File: skills/bypass-review/bypass-review [hunk 6]
```diff
@@ -648,6 +682,244 @@ def _render_error_events(lines: list[str], error_events: list[dict]) -> None:
...
+        instrumentable: set[str] = set()
+        uninstrumentable: set[str] = set()
+        for hook in manifest.get("hooks", []):
+            if not isinstance(hook, dict) or not hook.get("name"):
+                continue
+            body = hook.get("body") or ""
+            if isinstance(body, str) and body.endswith(".sh"):
+                uninstrumentable.add(hook["name"])
+            else:
+                instrumentable.add(hook["name"])
```
[the remark this answers is not in the document]
Comment (devseunggwan) [reply skills/bypass-review/bypass-review]: Fixed in e803c84 — added isinstance(name, str) guards to roster_split() and bash_group_roster() so a non-string `name` (e.g. {"name": 1}) is skipped before entering the roster, preventing the later sorted() TypeError. Regression test added (test_roster_split_skips_non_string_name).
